EIPS/eip-6963.md

```diff
@@ -0,0 +1,121 @@
+author: Pedro Gomes (@pedrouid), Kosala Hemachandra (@kvhnuke), Richard Moore (@ricmoo), Gregory Markou (@GregTheGreek)
+discussions-to: https://ethereum-magicians.org/t/eip-6963-multi-injected-provider-interface-aka-mipi
+status: Draft
+type: Standards Track
+category: Interface
+created: 2023-05-01
+requires: 1193
+replaces: 5749
+---
+
+## Abstract
+
+A Javascript interface injected by browser extensions installed by the user that enables user selection from multiple wallet providers using a window object labelled `window.evemproviders` instead of overwriting `window.ethereum` object.
+
+## Motivation
+
+Currently, wallet providers that offer browser extensions must inject their Ethereum providers (EIP-1193) into the same window object `window.ethereum` however this creates conflicts for users that may install more than one browser extension.
+
+Browser extensions are loaded in the web page and do not have a predictable pattern resulting in a race condition where the user is not in control to choose which wallet provider must be selected for exposing Ehtereum interface under the `window.ethereum` object.
+
+This results not only in a degraded user experience but also increases the barrier to entry for new browser extensions as users are forced to only install one browser extension at a time.
+
+Some browser extensions attempt to counteract this problem by delaying their injection to overwrite the same `window.ethereum` object which creates an unfair competition for wallet providers and lack of interoperability.
+
+In this proposal, we provide a solution that optimitizes for interoperability and enables fairer competition by lowering the barrier to entry for new wallet providers to improve the user experience on Ethereum.
+
+This is achieved by replacing the current window object `window.ethereum` with a new window object `window.evmproviders` that enables multiple browser extensions to be exposed for user selection in the web page.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Provider Info
+
+Each wallet provider will be exposed with the following interface `ProviderInfo` that will be used to display to the user:
+
+```typescript
+/**
+ * Represents the assets needed to display a wallet
+ */
+interface ProviderInfo {
+  uuid: string;
+  walletId: string;
+  name: string;
+  icon: string;
+  description?: string;
+}
+```
+
+The values in the `ProviderInfo` MUST be used as follows:
+
+- uuid - locally unique of the wallet provider (UUIDv4 compliant)
+- walletId - globally unique identifier of the walelt provider (eg. `my-wallet-extension` or `awesomewallet`)
+- name - human-readable name of the wallet provider (e.g. `MyWalletExtension` or `AwesomeWallet`)
+- icon - base64 encoded image (RFC-4648 complaint)
```

Comment:
It's definitely a must... otherwise we are going to incentivize centralization of some from of wallet registry

WalletConnect protocol also requires wallets and dapps to include an icon to avoid exactly this

It might seem irrelevant for low-level libraries like web3.js, ethers.js and viem.js but the end goal of this standard is to display options to the user which is going to be important for libraries like Web3Modal and others

Reply:
That's fair enough 😆 it should be any uri path

Reply:
I don't think it should be any uri, due to privacy concerns, we don't want users to be tracked by an image url. Image should be locally stored and best way to make sure that's the case if by making it base64

Reply:
Could we just standardize on an svg for icons? No tracking, no hosting, no url handling, just code.

Reply:
> Could we just standardize on an svg for icons? No tracking, no hosting, no url handling, just svg code.

What if the svg has an `<image>` element? 😉

This is why I believe this should be part of another EIP. There are so many nuances with icons. We would probably need to standardize this attribute further to avoid misuse, and might also be reasonable to add a standard to define what constitutes of a “default” design spec for the icon (ie. it would be weird for one icon to have rounded corners, and the others do not).

Reply:
> This is why I believe this should be part of another EIP. 
Can another EIP still be referenced in this one (6963)? To @pedrouid 's point, not having it will incentivise a centralised registry.

Reply:
I wouldn't say a this centralized registry of _just_ icons would necessarily be a bad thing until we have a proper standard for icons? Some form of centralization will probably exist regardless as RainbowKit and ConnectKit have [differing](https://github.com/rainbow-me/rainbowkit/pull/1166#discussion_r1169440416) [requirements](https://github.com/rainbow-me/rainbowkit/pull/1041#discussion_r1164381765) for their icons.

cc @DanielSinclair @lochie @0xAsimetriq

Reply:
I have no strong opinions either way. ConnectKit and other libraries could always just ignore these icons or just use them as fallbacks. But logos/icons can get pretty complex especially when considering width/height/ratio/contrast/rounded/etc. as context is important when using branding assets.
If this were to include icons/logos then we'd probably also need to include branding colors, which leads down a rabbit hole...

Correct me if I'm wrong, but image-tracking is probably a non-issue since if the provider exists in your `window.ethereum` object (or in this case `window.evmproviders`) then they'd be able to track in many other ways, so the svg/image uri would be the least of your worries.

Overall I think it's a safer bet to _not_ include icons, but also is there any harm to include them?

Reply:
I similarly have no strong opinion on icons, as lochie mentioned, we'd likely be overriding these in Web3Modal anyway. I do appreciate simplicity though, so I don't see icons as necessary. What's important imho is:

1. Structure that allows to access multiple providers
2. UUID's for each

Reply:
The point of standards to bring consensus over multiple parties

Honestly I think we are nitpicking too much on this topic

We should just agree that icons should be URIs, examples:

1.   data:image/png;base64, iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg==
2. ipfs://QmQqzMTavQgT4f4T5v6PWBp7XNKtoPmC9jvn12WPT3gkSE
3. https://ethereum.org/static/f541df14fca86543040c113725b5bd1a/99bcf/metamask.webp

URIs should be pretty much non-controversial... each wallet can decide which type of URI to use